### PR TITLE
Bugfix for smart indent issues #403 #404

### PR DESCRIPTION
**Pull request by HuanhuanSunMSFT**
@AndreSayreMSFT @EricJizbaMSFT @Microsoft/poshtools

### CODE REVIEW

File: PowerShellTools/LanguageService/SmartIndent.cs [hunk 1]
```diff
@@ -115,48 +115,51 @@ public SmartIndent(PowerShellLanguageInfo info, ITextView textView)
...
-		    indentation = IndentUtilities.GetCurrentLineIndentation(baselineText, tabSize);
-		}
+		return indentation;
 	    }
-	    else if (lastGroupStartEnd < lastGroupStartLine.End)
+	    if (lastGroupEndLine.LineNumber == line.LineNumber && String.IsNullOrWhiteSpace(textBetweenLineStartAndLastGroupEnd))
 	    {
-		string betweenText = textBuffer.CurrentSnapshot.GetText(lastGroupStartEnd, lastGroupStartLine.End - lastGroupStartEnd);
-		if (String.IsNullOrWhiteSpace(betweenText))
+		try
```
Comment (ASayre) [inline PowerShellTools/LanguageService/SmartIndent.cs]: In what cases do we expect this to fail?

File: PowerShellTools/LanguageService/SmartIndent.cs [hunk 2]
```diff
@@ -115,48 +115,51 @@ public SmartIndent(PowerShellLanguageInfo info, ITextView textView)
 
 	    int baselineEndPos = baseline.Extent.End.Position;
 	    var precedingGroupStarts = tokenSpans.FindAll(t => t.ClassificationType.IsOfType(Classifications.PowerShellGroupStart) && t.Start < baselineEndPos);
-	    var lastGroupStart = precedingGroupStarts.FindLast(p => startBraces[p.Start] >= baselineEndPos);
+	    var lastGroupStart = precedingGroupStarts.FindLast(p =>
+	    {
+		int closeBrace;
+		return !startBraces.TryGetValue(p.Start, out closeBrace) || closeBrace >= baselineEndPos;
```
Comment (ASayre) [inline PowerShellTools/LanguageService/SmartIndent.cs:121]: What cases do we expect this to fail?

Comment (HuanhuanSunMSFT) [inline PowerShellTools/LanguageService/SmartIndent.cs:121]: We only have Group start but not the paired braces.

**Comment by ASayre** [comment]
ping re: that try/catch?  If there's a crash that's causing it, we should fix that instead of hiding it...

**Comment by HuanhuanSunMSFT** [comment]
I cannot think of any cases that will throw exceptions there but cannot guarantee, so add the try/catch as a preventive.   @AndreSayreMSFT

**Comment by ASayre** [comment]
Per convo, looks good after try/catch; If we need to make a tactical fix because of any errors, let's do so upon discovering the errors!

**Pull request merged**